File: src/tools/runCode.ts
// run_code — execute LLM-written JS in the L4 sandbox (child process,
// fs locked to job dir). Available ONLY to sub-agents, never the main agent.

import { Tool, ToolResult } from '../core/types.js'
import { runInSandbox } from '../sandbox/runner.js'

export const runCodeTool: Tool = {
  name: 'run_code',
  description:
    'Run JavaScript (ESM, Node) in an isolated sandbox. Filesystem access is limited to the job directory (cwd). Input files you list are copied into cwd first. Use console.log for output; write any produced files to cwd. Returns stdout, stderr, and paths of files created.',
  inputSchema: {
    type: 'object',
    properties: {
      code: { type: 'string', description: 'JavaScript module code to run' },
      inputFiles: {
        type: 'array',
        items: { type: 'string' },
        description: 'Absolute paths of files to copy into the sandbox cwd before running',
      },
    },
    required: ['code'],
  },
  async execute(input): Promise<ToolResult> {
    try {
      const result = await runInSandbox(
        String(input.code),
        Array.isArray(input.inputFiles) ? input.inputFiles.map(String) : []
      )
      return {
        success: result.success,
        output: {
          stdout: result.stdout,
          stderr: result.stderr,
          filesCreated: result.files,
        },
        error: result.success ? undefined : `exit non-zero. stderr: ${result.stderr.slice(0, 500)}`,
      }
    } catch (e) {
      return { success: false, output: null, error: String(e) }
    }
  },
}

File: python/app/tools/run_code.py
"""run_code — execute LLM-written Python in the L4 sandbox. Available ONLY to
sub-agents (least privilege), never the main agent."""

from langchain_core.tools import tool

from ..sandbox.runner import run_in_sandbox


@tool
def run_code(code: str, input_files: list[str] | None = None) -> dict:
    """Run Python code in an isolated sandbox. Working directory is a private job dir; input files you list are copied into it first. Use print() for output; write any produced files to the working directory. Returns stdout, stderr, and paths of files created."""
    try:
        result = run_in_sandbox(code, input_files or [])
        out = {
            "success": result.success,
            "output": {
                "stdout": result.stdout,
                "stderr": result.stderr,
                "filesCreated": result.files,
            },
        }
        if not result.success:
            out["error"] = f"exit non-zero. stderr: {result.stderr[:500]}"
        return out
    except Exception as e:
        return {"success": False, "error": str(e)}
